### PR TITLE
WIP: Add require-valid-categories rule

### PR DESCRIPTION
**Pull request by macklinu**
See #60 

This is a start at implementing the rule in #60. Still a WIP, but wanted to open the PR to start a discussion and also let you know that someone is working on it. 🎉 

TODO: 

- [ ] add rule category
- [ ] add proper JSDoc
- [ ] add more test cases (old rule format, new rule format, empty category option, no category option, invalid category option, like numbers)
- [ ] add `docs/rules/require-valid-categories.md`

### CODE REVIEW

File: lib/rules/require-valid-categories.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const utils = require('../utils');
+
+module.exports = {
+  meta: {
+    docs: {},
+    category: '', // fixme: add valid category
+    schema: {
```
Comment (macklinu) [inline lib/rules/require-valid-categories.js]: My thought is that this rule should be opt-in only and requires options (otherwise what categories would you validate against)? As such, the only option this rule takes is an array of strings containing at least one value. Otherwise, ESLint should report that the rule was configured improperly. Thoughts on this?

File: lib/rules/require-valid-categories.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          docsPropNode.value.properties &&
+          docsPropNode.value.properties.find(p => p.type === 'Property' && utils.getKeyName(p) === 'category');
+
+        if (isValidCategory(categoryPropNode && categoryPropNode.value)) {
+          return;
+        }
+
+        context.report({
+          node: categoryPropNode.value,
+          message: 'Not a valid category.',
```
Comment (macklinu) [inline lib/rules/require-valid-categories.js]: Probably could use a better message too. 🤔

File: lib/rules/require-valid-categories.js (hunk 1 above)
Comment (not-an-aardvark) [reply lib/rules/require-valid-categories.js]: My only concern here is that this plugin has the `plugin:eslint-plugin/all` config, which currently claims to enable all rules from the plugin. This isn't possible unless all rules can be run without providing an explicit configuration. So my slight preference would be to have some default set of categories (e.g. `Possible Errors`, `Best Practices`, `Stylistic Issues`) if the user doesn't provide any categories on their own.

File: lib/rules/require-valid-categories.js (hunk 2 above)
Comment (not-an-aardvark) [reply lib/rules/require-valid-categories.js]: Agreed -- maybe add the names of the valid categories? For example: `Category should be one of 'Foo', 'Bar', or 'Baz'.`

**Comment by aladdin-add** [comment]
@armano2  the rule was configurable,  as described in the issue #60 😄 
> This rule would require that all lint rules have a meta.docs.category property which is one of several user-specified strings. 

friendly ping @macklinu are you still working on this?

**Comment by macklinu** [comment]
@aladdin-add sorry for dropping the ball on this (missed the message in my email inbox). Will pick back up on this rule this week. 😄

File: lib/rules/require-valid-categories.js (hunk 1 above)
Comment (j-f1) [reply lib/rules/require-valid-categories.js]: Perhaps all the categories ESLint core uses? It would be nice to be able to use those and combine them with plugin-specific options.

**Comment by aladdin-add** [comment]
friendly ping @macklinu

**Comment by aladdin-add** [comment]
closing, as the `category` will be removed in eslint v8. :)
